test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+""" Test node eviction logic
+
+When the number of peers has reached the limit of maximum connections,
+the next connecting inbound peer will trigger the eviction mechanism.
+We cannot currently test the parts of the eviction logic that are based on
+address/netgroup since in the current framework, all peers are connecting from
+the same local address. See Issue #14210 for more info.
+
+Therefore, this test is limited to the remaining protection criteria:
+1.) Protect 4 nodes having sent us a block most recently
+2.) Protect 4 nodes having sent us a transaction most recently
+3.) Protect 8 nodes with the smallest minimum ping
+"""
+
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mininode import P2PInterface, P2PDataStore
+from test_framework.util import assert_equal, wait_until
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
+
+
+class SlowP2PDataStore(P2PDataStore):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+class SlowP2PInterface(P2PInterface):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+
+class P2PEvict(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        # The choice of 32 maxconnections results in a maximum of 21 inbound connections
+        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
+        self.extra_args = [['-maxconnections=32']]
+
+    def setup_network(self):
+        self.setup_nodes()
+
```

Comment:
Unused code?
```diff
-    def setup_network(self):
-        self.setup_nodes()
-
```

Reply:
or better yet, keep it per suggestion in https://github.com/jonatack/bitcoin/commit/c9e96a1 and add:
```diff
    def run_test(self):
+        self.log.info("Set up test peer network")
+        self.setup_test_peers()
```
This provides the important advantage of feedback on how much time is spent in setup while running the test.

Reply:
The `setup_network` override is not needed imo, removed - thanks!

I added several logging messages for the initial phase (as also suggested below).
 
I am still a bit hesitant with respect to the setup network / test separation, because I don't really view most of it as network setup - in the network setup phase, one node is created, and then in the second phase, we simulate inbound peers connecting to it, sending blocks and transactions, until all slots are taken and a peer gets evicted.
But I will change it if others also think that this is preferable!

Reply:
I'm unsure but suspect one would see `connect_node` statements in the tests if the override was needed.

If you prefer to keep the setup and the assertions all in one block, then no need for `self.protected_peers` and `self.evicted_peers`; that was done in order to separate the network setup from the assertions.

Reply:
Done, made them local again.